### PR TITLE
set err to val with get response status non-nil Fixes #56

### LINKED ISSUE #56: status code is ignored
From section 4.1(introduction) of [the spec](https://code.google.com/p/memcached/wiki/MemcacheBinaryProtocol):

```
If the status code of a response packet is non-nil,
   the body of the packet will contain a textual error message.
```

From section 4.2.1 (get request example):

```
 If the item exist on the server the following packet is returned,
   otherwise a packet with status code != 0 will be returned (see
   Introduction (Section 4.1))
```

When I set a non-nil status code on my C++ server, from a get to a non existing key, I see the client's 'get' callback invoked with null error and null value.  The callback should be invoked with error set to the response body and null for the value.

### PR DESCRIPTION
**Pull request by nickdesaulniers**

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
just pushed fixes for Travis.  Looks like GH is having issues picking up changes to a branch today...

**Comment by nickdesaulniers** [comment]
just got a confirmation email from GH support that something's up:
```
Hi Nick,

Sorry for the trouble. We made a change that accidentally resulted in some PR bugginess.

We're working on a fix, and I'll let you know when we have an update.

Cheers,
Jamie
```

File: lib/memjs/memjs.js [hunk 1]
```diff
@@ -115,7 +115,7 @@ Client.prototype.get = function(key, callback) {
       callback && callback(null, response.val, response.extras)
       break;
     case 1:
-      callback && callback(null, null, null);
+      callback && callback(response.val.toString(), null, null);
```
Comment (alevy) [inline lib/memjs/memjs.js]: For consistency, I think we should make this:

```
callback && callback(new Error(response.val.toString(), null, null);
```

**Comment by alevy** [comment]
Do you have a test for this regression, by chance? Otherwise I can add one.

**Comment by nickdesaulniers** [comment]
I will update this test: https://github.com/alevy/memjs/blob/master/test/client_test.js#L31-L54

**Comment by nickdesaulniers** [comment]
review? @alevy

**Comment by alevy** [comment]
@nickdesaulniers thanks for updating the pull request!

On second look, I'm a bit concerned with how a user differentiates between a benign, "key not found", error and a more problematic error case, like a server error. With how the released callback works, you would simply:

```
client.get('key', function(err, val) {
  if (err) {
    // Bad error, e.g. server is down
  } else if (val == null) {
    // No error, just key not found
  } else {
    // key found
  }
}
``` 
I'm not sure, with this code, how you handle errors in a sane way...

This might be salvageable, though. The specific error code `0x1`, always means "key not found", so maybe that's not where we need to change the callback parameters. In the default case, though, perhaps the error message should use `response.val`, as in your pull request, instead of `response.header.status`.

Sorry to give you the run around, I should have thought more carefully before. If we can convince ourselves this is the right API, I'm happy merging as is.

**Comment by nickdesaulniers** [comment]
> The specific error code 0x1, always means "key not found"

That's not necessarily the case, from the spec section 4.2.1 (get request example):
```
 If the item exist on the server the following packet is returned,
   otherwise a packet with status code != 0 will be returned (see
   Introduction (Section 4.1))
```

To support the use case you describe, shouldn't the client have an on disconnect callback that users can assign?

Besides, if the server is down, the client will try twice and if not throw an error.  Let me whip up an example in code...brb

**Comment by alevy** [comment]
Well... not exactly. First, the `get` function acts on a cluster of memcache servers, not a single one, so a disconnect callback seems a bit odd (since you probably haven't disconnected from the whole cluster). Second, if you look at the response code section in the spec (https://code.google.com/p/memcached/wiki/BinaryProtocolRevamped#Response_Status), you'll see that not found is always status code `0x1`. There are other status codes (such as "busy"), which means different things, and map to a different class of error case.

**Comment by nickdesaulniers** [comment]
This seems to work as expected with my patch:
```coffeescript
memjs = require 'memjs'
client = memjs.Client.create 'localhost:3000'

setTimeout ->
  client.get 'hello', (err, val) ->
    if err
      if err.code and err.code is 'ECONNREFUSED'
        console.log 'server down'
      else
        console.log err.message
    else
      console.log val
, 5000
```

**Comment by alevy** [comment]
In general, I believe this is how memcache clients handle things --- not found keys are returned as null-like values, while other errors throw exceptions (which I think most closely maps to an error parameters in the callback for Node). Of course, just because other clients do that doesn't mean it's right :)

**Comment by alevy** [comment]
Yes, that works for cases like a failed server, but not for other eroneous status codes, like 'busy'. You also probably want to do something more useful, in general, when the value is not found. A common case, for example, is to check if something is in the cache, and if not recreate it and set it to the cache:

```
client.get('expensivequery', function(err, val) {
  if (err) {
    // potentially do something catestrphic like send a 500 error to user, queue request if busy, restart app server, etc
  } else if (val == null) {
     db.query("select * from posts where [very-expensive-where-clause]", function(val) {
        client.set('expensivequery', val);
        resp.render(val);
     });
  } else {
    resp.render(val);
  }
}
```

The way do it with your suggested changes would be for us to do something like set the error code to the response status and for the user to check if that `===1`. That might be fine, but it leaves two questions to answer:

1. Is the interface actually better at that point? Particularly, what's the value of having the string message "NOT FOUND" if you can otherwise determine that the value was not found.

2. How much existing user code does it break and is it worth breaking that code?

I don't think I'll have much to add after this, so if we're stuck not convincing each other at this point, let's try solicit one or two other opinions.

**Comment by dterei** [comment]
I think we shouldn't make this change.

1. Not found isn't an error, it's an expected result, memcache is fairly careful to refer to the field as a `status` field since it uses it for both true errors and null results.

2. The error message returned isn't more descriptive than how we currently communicate Not Found. Perhaps if you implemented a custom memcache server this would be different but that's complicating the general use case for a rare case.

3. The patch only makes the change for `GET` and only for Key NotFound. If we do switch to the design proposed in the patch we should do it universally.

Basically, while I appreciate the contribution, I don't think it's  change we want.
